### PR TITLE
[ticket/14972] PHP 7.2 compatibility

### PR DESCRIPTION
**Pull request by rxu**
Checklist:

- [ ] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [ ] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

<a href="https://tracker.phpbb.com/browse/PHPBB3-14972">PHPBB3-14972</a>.

### CODE REVIEW

**Comment by rxu** [comment]
!set WIP

**Comment by VSEphpbb** [comment]
Looks like `each()` is deprecated in 7.2

File: phpBB/phpbb/db/driver/driver.php [hunk 1]
```diff
@@ -537,6 +537,11 @@ function sql_build_array($query, $assoc_ary = false)
 	*/
 	function sql_in_set($field, $array, $negate = false, $allow_empty_set = false)
 	{
+		if (!is_array($array))
+		{
+			$array = array($array);
+		}
```
Comment (Nicofuma) [inline phpBB/phpbb/db/driver/driver.php]: you can replace it by `$array = (array) $array;` which is simpler imo

**Comment by rxu** [comment]
Regrettably, the ```s9e/TextFormatter``` is not 7.2-dev compatible currently, so the Textformatter-related failures fixing is postponed until it got this compatible. tested it here https://travis-ci.org/rxu/TextFormatter/jobs/229649094 @JoshyPHP

**Comment by JoshyPHP** [comment]
@rxu I'll look into it but don't expect anything to happen before PHP releases their first beta.

**Comment by rxu** [comment]
Twig is also incompatible with PHP 7.2-dev currently, so I give up for now (I guess, at least until 3rd-party components such as Textformatter and Twig got updated to comply PHP 7.2 requirements).

**Comment by rxu** [comment]
It seems that the Textformatter is already adjusted to make PHP 7.2 tests green 😛

**Comment by rxu** [comment]
@JoshyPHP Thanks for updating the Textformatter, and could  you give an idea what would cause this error https://travis-ci.org/phpbb/phpbb/jobs/230667648#L813

**Comment by JoshyPHP** [comment]
@rxu You're welcome. I don't have a local install though so it'll have to remain an allowed failure on my side until there's a downloadable beta.

I looked at this error message and I suppose there's something that prevents `--{E` from being escaped to `--{{E` in `phpbb\textformatter\s9e\factory::escape_html_attribute()`.

**Comment by rxu** [comment]
Yeah, just doing some investigations.

File: phpBB/includes/functions.php [hunk 2]
```diff
@@ -590,7 +590,7 @@ function markread($mode, $forum_id = false, $topic_id = false, $post_time = 0, $
 
 	if ($mode == 'all')
 	{
-		if ($forum_id === false || !sizeof($forum_id))
+		if ($forum_id === false || empty($forum_id))
```
Comment (VSEphpbb) [inline phpBB/includes/functions.php]: wouldn't just `empty($forum_id)` suffice?

Comment (rxu) [inline phpBB/includes/functions.php]: Probably it would, $forum_id is assumed to be ```false```, ```int``` or ```array``` here.

**Comment by JoshyPHP** [comment]
@rxu The test failing on `--{{E` should be fixed by https://github.com/s9e/TextFormatter/commit/ed780e3b9cd9b346c49cde153254989fe6170eff.

**Comment by rxu** [comment]
@JoshyPHP Thanks again, with 0.9.6 textformatter tests are green now.

**Comment by rxu** [comment]
Now trying to sort out Twig issues https://github.com/twigphp/Twig/issues/2464

File: phpBB/phpbb/template/context.php [hunk 3]
```diff
@@ -282,39 +282,26 @@ public function find_key_index($blockname, $key)
...
-				$index = sizeof($block[$name]) - 1;
-			}
 			if (!isset($block[$name]))
 			{
 				return false;
 			}
-			$block = $block[$name];
-			if (!isset($block[$index]))
+
+			$index = (strpos($blocks[$i], '[]') === $pos) ? (count($block[$name]) - 1) : (min((int) substr($blocks[$i], $pos + 1, -1), count($block[$name]) - 1));
```
Comment (javiexin) [inline phpBB/phpbb/template/context.php]: $index = (!$pos || strpos($blocks[$i], '[]') === $pos) ? (count($block[$name]) - 1) : (min((int) substr($blocks[$i], $pos + 1, -1), count($block[$name]) - 1));

**Comment by rxu** [comment]
OK, now after the ```guzzle/guzzle``` 6.3.0 release we got all upgraded components in this PR (Twig, Textformatter and Guzzle) PHP 7.2-compatible. So we should be ready to make a clean version of this PR for master (which could be then probably backported to 3.2.x).

**Comment by rxu** [comment]
Replaced by #4862.

**Pull request closed without merging**